code/modules/chemistry/Reagents-Drugs.dm

```diff
@@ -492,6 +492,127 @@ datum
 				..()
 				return
 
+		drug/caffeine        //Unified chem for lots of caffeinated drinks, similar to how ethanol functions
+			name = "caffeine"
+			id = "caffeine"
+			description = "An addictive stimulant contained in coffee beans and many caffeinated beverages."
+			reagent_state = LIQUID
+			fluid_r = 230
+			fluid_g = 220
+			fluid_b = 230
+			addiction_prob = 1 //It lasts a while, it's not as low as it seems
+			addiction_min = 100
+			max_addiction_severity = "LOW"
+			depletion_rate = 0.1
+			taste = "bitter"
+			overdose = 60
+			threshold = THRESHOLD_INIT
+			var/heart_failure_counter = 0
+
+			cross_threshold_over()
+				if(ismob(holder?.my_atom))
+					var/mob/M = holder.my_atom
+					APPLY_ATOM_PROPERTY(M, PROP_MOB_STAMINA_REGEN_BONUS, "caffeine_rush", 3)
+				..()
+
+			cross_threshold_under()
+				if(ismob(holder?.my_atom))
+					var/mob/M = holder.my_atom
+					REMOVE_ATOM_PROPERTY(M, PROP_MOB_STAMINA_REGEN_BONUS, "caffeine_rush")
+				..()
+
+			on_mob_life(var/mob/M, var/mult = 1)
+				if(!M) M = holder.my_atom
+				var/caffeine_amt = holder.get_reagent_amount(src.id)
+				if (M.get_eye_blurry()) //reducess vision blurryness
+					M.change_eye_blurry(-1 * mult)
+				if (M.bodytemperature < M.base_body_temp) // So it doesn't act like supermint
+					M.bodytemperature = min(M.base_body_temp, M.bodytemperature+(5 * mult))
+				if (heart_failure_counter > 150 && ishuman(M)) // This has to get pretty high for bad things to happen
+					var/mob/living/L = M
+					L.contract_disease(/datum/ailment/malady/heartfailure, null, null, 1)
+					heart_failure_counter = 0
+
+				if (caffeine_amt > 30) //Large amounts don't last as much as smaller trace amounts
+					depletion_rate = 0.2
+					M.sleeping = 0 //Causes insomnia
+				else
+					depletion_rate = 0.1
+
+				switch(caffeine_amt)
+					if(0 to 5)   //This is your trace amount of caffeine, doesn't do much
+						stun_resist = 3
+
+					if(5 to 20)  //A regular coffe mug's worth
+						if (M.get_eye_blurry() && prob(75))
+							M.change_eye_blurry(-1 * mult)
+						stun_resist = 6
```

Comment:
setting stun resist like this does not work

Reply:
Oh, i see, alrighty, gonna figure out a way to make it work then.

Reply:
This seems to work pretty well, but if it's still too complex, i can keep it at a set value?